File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.py
def solution(board, skill):
    answer = 0
    N = len(board)
    M = len(board[0])
    acc_sum = [[0] * (M+1) for _ in range(N+1)]
    for cmd in skill:
        skill_type, start_r, start_c, end_r, end_c, amount = cmd
        if skill_type == 1:
            amount *= -1 
        acc_sum[start_r][start_c] += amount
        acc_sum[start_r][end_c + 1] -= amount
        acc_sum[end_r + 1][end_c + 1] += amount
        acc_sum[end_r + 1][start_c] -= amount
    
    
    for i in range(N):
        for j in range(1, M):
            acc_sum[i][j] += acc_sum[i][j-1]
            
    for i in range(1, N):
        for j in range(M):
            acc_sum[i][j] += acc_sum[i-1][j]
            
    for i in range(N):
        for j in range(M):
            if board[i][j] + acc_sum[i][j] >= 1:
                answer += 1
    
    return answer

File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(vector<vector<int>> board, vector<vector<int>> skill) {
    int answer = 0;
    vector<vector<int>> acc_sum = vector<vector<int>>(board.size() + 1, vector<int>(board[0].size() + 1, 0));
    
    
    for (vector<int> cmd: skill) {
        int type = cmd[0];
        int r1 = cmd[1];
        int c1 = cmd[2];
        int r2 = cmd[3]; 
        int c2 = cmd[4];
        int degree = cmd[5];

        if (type == 1){
            acc_sum[r1][c1] -= degree;
            acc_sum[r1][c2 + 1] += degree;
            acc_sum[r2+1][c1] += degree;
            acc_sum[r2+1][c2+1] -= degree;
        } else {
            acc_sum[r1][c1] += degree;
            acc_sum[r1][c2 + 1] -= degree;
            acc_sum[r2+1][c1] -= degree;
            acc_sum[r2+1][c2 + 1] += degree;
        }
    }
    
    for (int i = 0; i < acc_sum.size(); i++) {
        for (int j = 1; j < acc_sum[0].size(); j++) {
            acc_sum[i][j] += acc_sum[i][j-1];
        }
    }
    
    for (int j = 0; j < acc_sum[0].size(); j++) {
        for (int i = 1; i < acc_sum.size(); i++) {
            acc_sum[i][j] += acc_sum[i-1][j];
        }
    }
    
    for (int i = 0; i < board.size(); i++) {
        for (int j = 0; j < board[0].size(); j++) {
            
            if (board[i][j] + acc_sum[i][j] >= 1) {
                answer += 1;
            }
        }
    }
    
    
    
    
    
    
    
    
    return answer;
}